src/farming/library.cairo

```diff
@@ -0,0 +1,1039 @@
+
+@storage_var
+func CarbonableFarming_total_registered_time_() -> (time: felt) {
+}
+
+// Contract user records
+
+@storage_var
+func CarbonableFarming_absorption_(address: felt) -> (absorption: felt) {
+}
+
+@storage_var
+func CarbonableFarming_sale_(address: felt) -> (sale: felt) {
+}
+
+@storage_var
+func CarbonableFarming_registered_value_(address: felt) -> (value: Uint256) {
+}
+
+@storage_var
+func CarbonableFarming_registered_time_(address: felt) -> (time: felt) {
+}
+
+namespace CarbonableFarming {
+    //
+    // Constructor
+    //
+
+    func initializer{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        carbonable_project_address: felt, carbonable_project_slot: Uint256
+    ) {
+        CarbonableFarming_carbonable_project_address_.write(carbonable_project_address);
+        CarbonableFarming_carbonable_project_slot_.write(carbonable_project_slot);
+        return ();
+    }
+
+    //
+    // Getters
+    //
+
+    func carbonable_project_address{
+        syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr
+    }() -> (carbonable_project_address: felt) {
+        let (carbonable_project_address) = CarbonableFarming_carbonable_project_address_.read();
+        return (carbonable_project_address=carbonable_project_address);
+    }
+
+    func carbonable_project_slot{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        ) -> (carbonable_project_slot: Uint256) {
+        let (carbonable_project_slot) = CarbonableFarming_carbonable_project_slot_.read();
+        return (carbonable_project_slot=carbonable_project_slot);
+    }
+
+    func total_deposited{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        value: Uint256
+    ) {
+        let (value) = CarbonableFarming_total_registered_value_.read();
+        return (value=value);
+    }
+
+    func total_absorption{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        absorption: felt
+    ) {
+        let (computed) = _compute_total_absorption();
+        let (stored) = CarbonableFarming_total_absorption_.read();
+        return (absorption=computed + stored);
+    }
+
+    func max_absorption{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        absorption: felt
+    ) {
+        // [Compute] Max absorption, which is the whole project absorption until now
+        let (contract_address) = CarbonableFarming_carbonable_project_address_.read();
+        let (slot) = CarbonableFarming_carbonable_project_slot_.read();
+        let (absorption) = ICarbonableProject.getCurrentAbsorption(
+            contract_address=contract_address, slot=slot
+        );
+        return (absorption=absorption);
+    }
+
+    func total_sale{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        sale: felt
+    ) {
+        alloc_locals;
+
+        // [Check] Price is setup, otherwise return 0
+        let (len) = Array.read_len(key=TIME_SK);
```

Comment:
If I follow the comment, the check should be `key=PRICE_SK` instead of `key=TIME_SK` ?

Reply:
Price in comment means Price curve which is defined with 2 arrays `times`  and `prices` (basically x and y values).
All checks are setup to ensure you cannot have a 2 different sizes for both arrays, so checking `times`  len is litterally the same thing than checking `prices` array.
So nothing to change imo.